test/malloc_compliance_tests.cpp

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// Basic tests
+// ISO/IEC 9899:201x 7.22.3 compliance
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "malloc_compliance_tests.hpp"
+#include "test_helpers.h"
+#include "umf/memory_pool.h"
+
+#include "base.hpp"
+using umf_test::test;
+
+//------------------------------------------------------------------------
+// Configurable defs
+//------------------------------------------------------------------------
+
+#define KB 1024
+#define MB (1024 * KB)
+
+#define MAX_ALLOC_SIZE (1 * MB)
+#define ITERATIONS 100
+
+//------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------
+
+#define ASSERT_MSG(expr, msg)                                                  \
+    do {                                                                       \
+        if (!(expr)) {                                                         \
+            DEBUG_MSG(expr, msg);                                              \
+            ASSERT_NE((intptr_t)(expr), 0);                                    \
+        }                                                                      \
+    } while (0)
+
+static inline size_t rand_alloc_size(int max) { return rand() % max; }
+
+static inline void free_memory(umf_memory_pool_handle_t hPool,
+                               void *ptr[ITERATIONS]) {
+    for (int i = 0; i < ITERATIONS; i++) {
+        umfPoolFree(hPool, ptr[i]);
+    }
+}
+
+//------------------------------------------------------------------------
+// Tests
+// -----------------------------------------------------------------------
+
+// ISO/IEC 9899:TC3 7.20.3.3
+void malloc_compliance_test(umf_memory_pool_handle_t hPool) {
+    void *alloc_ptr[ITERATIONS];
+    size_t alloc_ptr_size[ITERATIONS];
+
+    // Each allocation shall yield a pointer to an object disjoint from
+    // any other object.
+    for (int i = 0; i < ITERATIONS; i++) {
+        alloc_ptr_size[i] = rand_alloc_size(MAX_ALLOC_SIZE);
+        alloc_ptr[i] = umfPoolMalloc(hPool, alloc_ptr_size[i]);
+
+        // Supported under USE_WEAK_ALIGNMENT_RULES macro
+        // For compatibility, on 64-bit systems malloc should align to 16 bytes
+        size_t alignment =
+            (sizeof(intptr_t) > 4 && alloc_ptr_size[i] > 8) ? 16 : 8;
+        ASSERT_MSG(
+            is_aligned(alloc_ptr[i], alignment),
+            "Malloc should return pointer that is suitably aligned so that "
+            "it may be assigned to a pointer to any type of object ");
+
+        // Fill memory with zeros
+        ASSERT_MSG(alloc_ptr[i],
+                   "malloc returned NULL, couldn't allocate much memory");
+        memset(alloc_ptr[i], 0, alloc_ptr_size[i]);
+    }
+    for (int i = 0; i < ITERATIONS; i++) {
+        ASSERT_MSG(is_zero_mem(alloc_ptr[i], alloc_ptr_size[i]),
```

Comment:
is this a reliable way to check this? wouldn't it be better to fill the buffer with a different pattern for each allocation (e.g., the id of the allocation, and then cap the iters at 256)?

Reply:
Done